Files.list behavior differs on non-directories between JVM and Scala Native

**Scala Native version**: 0.5.8
**Platform**: MacOS. Darwin 24.6.0 arm64

Hey everyone,
I encountered an issue while working on https://github.com/typelevel/fs2/pull/3586.

The issue occurs when `Files.list` is called on a path that isn’t a directory.
On JVM `Files.list` throws a `NotDirectoryException`, while on Scala Native, that same exception is wrapped in an `UncheckedIOException`.

Here's a minimal example to reproduce the issue:
```scala
//> using scala "2.13.16"

import java.nio.file.Path
import java.nio.file.{Files, Paths}

object Main {
  def main(args: Array[String]): Unit = {
    val dir = Paths.get("./test.scala") // it's a file, not a directory
    val fileStream = Files.list(dir)
    try {
      val it = fileStream.iterator()
      val files = new java.util.HashSet[Path]()
      while (it.hasNext()) // exhaust the iterator
        files.add(it.next())
    } finally fileStream.close()
  }
}
```

JVM:
```
$ scala-cli  test.scala
Compiling project (Scala 2.13.16, JVM (21))
Compiled project (Scala 2.13.16, JVM (21))
Exception in thread "main" java.nio.file.NotDirectoryException: ./test.scala
        at java.base/sun.nio.fs.UnixFileSystemProvider.newDirectoryStream(UnixFileSystemProvider.java:460)
        at java.base/java.nio.file.Files.newDirectoryStream(Files.java:481)
        at java.base/java.nio.file.Files.list(Files.java:3772)
        at Main$.main(test.scala:9)
        at Main.main(test.scala)
```

Native:
```
$ scala-cli --native test.scala                                                                                                                                                                       on  main [!?]
Compiling project (Scala 2.13.16, Scala Native 0.5.8)
Compiled project (Scala 2.13.16, Scala Native 0.5.8)
[info] Linking (multithreadingEnabled=true, disable if not used) (1210 ms)
[info] Discovered 987 classes and 6511 methods after classloading
[info] Checking intermediate code (quick) (57 ms)
[info] Multithreading was not explicitly enabled - initial class loading has not detected any usage of system threads. Multithreading support will be disabled to improve performance.
[info] Linking (multithreadingEnabled=false) (748 ms)
[info] Discovered 868 classes and 5594 methods after classloading
[info] Checking intermediate code (quick) (23 ms)
[info] Discovered 849 classes and 4298 methods after optimization
[info] Optimizing (debug mode) (1085 ms)
[info] Produced 8 LLVM IR files
[info] Generating intermediate code (1040 ms)
[info] Compiling to native code (1662 ms)
[info] Linking with [pthread, dl, m]
[info] Linking native code (immix gc, none lto) (164 ms)
[info] Postprocessing (0 ms)
[info] Total (5769 ms)
Exception in thread "main" java.io.UncheckedIOException
        at java.nio.file.Files$PosixDir$1.$anonfun$posixDir$1(Unknown Source)
        at java.nio.file.Files$PosixDir$1$$Lambda$1.apply(Unknown Source)
        at scala.scalanative.unsafe.Zone$.acquire(Unknown Source)
        at java.nio.file.Files$PosixDir$1.posixDir$lzycompute(Unknown Source)
        at java.nio.file.Files$PosixDir$1$$anon$1.tryAdvance(Unknown Source)
        at java.util.Spliterators$$anon$8.hasNext(Unknown Source)
        at Main$.main(Unknown Source)
        at Main.main(Unknown Source)
        at <none>.main(Unknown Source)
Caused by: java.nio.file.NotDirectoryException: ./test.scala
        at java.nio.file.PosixException$.apply(Unknown Source)
        ... 9 more
```

--- a/javalib/src/main/scala/java/nio/file/Files.scala
+++ b/javalib/src/main/scala/java/nio/file/Files.scala
@@ -827,7 +827,7 @@ object Files {
       private val posixDir: Ptr[DIR] = Zone.acquire { implicit z =>
         val ptr = opendir(toCString(dirString))
         if (ptr == null)
-          throw new UncheckedIOException(PosixException(dirString, errno))
+          throw PosixException(dirString, errno)
 
         posixDirClosed = false
         ptr
@@ -862,9 +862,7 @@ object Files {
 
               if (entry == null) {
                 if (errno != 0) {
-                  throw new UncheckedIOException(
-                    PosixException(dirString, errno)
-                  )
+                  throw PosixException(dirString, errno)
                 } else { // End of OS directory stream
                   closeImpl()
                   false
@@ -905,7 +903,7 @@ object Files {
         if (!posixDirClosed) {
           val err = closedir(posixDir)
           if (err != 0)
-            throw new UncheckedIOException(PosixException(dirString, errno))
+            throw PosixException(dirString, errno)
 
           posixDirClosed = true
         }

--- a/unit-tests/shared/src/test/scala/org/scalanative/testsuite/javalib/nio/file/FilesTest.scala
+++ b/unit-tests/shared/src/test/scala/org/scalanative/testsuite/javalib/nio/file/FilesTest.scala
@@ -1016,6 +1016,26 @@ class FilesTest {
     }
   }
 
+  // Issue #4432
+  @Test def filesListMatchJvmExceptions(): Unit = {
+    withTemporaryDirectory { dirFile =>
+      val dir = dirFile.toPath()
+      val file = dir.resolve("fileNotDir")
+      Files.createFile(file)
+
+      assertTrue("a1", Files.exists(file) && Files.isRegularFile(file))
+
+      try {
+        assertThrows(
+          classOf[NotDirectoryException],
+          Files.list(file) // it's a file, not a directory
+        )
+      } finally {
+        Files.delete(file)
+      }
+    }
+  }
+
   @Test def filesReadSymbolicLinkCanReadValidSymbolicLink(): Unit = {
     assumeShouldTestSymlinks()
 

--- a/unit-tests/shared/src/test/scala/org/scalanative/testsuite/javalib/nio/file/FilesTest.scala
+++ b/unit-tests/shared/src/test/scala/org/scalanative/testsuite/javalib/nio/file/FilesTest.scala
@@ -1025,14 +1025,21 @@ class FilesTest {
 
       assertTrue("a1", Files.exists(file) && Files.isRegularFile(file))
 
-      try {
-        assertThrows(
-          classOf[NotDirectoryException],
-          Files.list(file) // it's a file, not a directory
-        )
-      } finally {
-        Files.delete(file)
-      }
+      val expectedException =
+        if (isWindows)
+
+          try {
+            val expectedException =
+              if (isWindows) classOf[IOException]
+              else classOf[NotDirectoryException]
+
+            assertThrows(
+              expectedException,
+              Files.list(file) // it's a file, not a directory
+            )
+          } finally {
+            Files.delete(file)
+          }
     }
   }
 

